File: monitor/src/main/java/com/caibowen/prma/monitor/Monitor.java
package com.caibowen.prma.monitor;

import com.caibowen.gplume.annotation.Const;
import com.caibowen.prma.api.model.EventVO;
import com.caibowen.prma.monitor.eval.Evaluator;
import com.caibowen.prma.monitor.notify.Notifier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nonnull;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Executor;

/**
 * @author BowenCai
 * @since 28/11/2014.
 */
public class Monitor {

    private static final Logger LOG = LoggerFactory.getLogger(Monitor.class);

    private Executor executor;

    protected Evaluator evaluator;
    protected CopyOnWriteArrayList<Notifier> notifiers;
    protected HashMap<String, Notifier> maps;

    public Monitor(){}

    public Monitor(Evaluator evaluator, List<Notifier> notifiers) {
        this.evaluator = evaluator;
        this.notifiers = new CopyOnWriteArrayList<>(notifiers);
        for (Notifier n : notifiers)
            maps.put(n.getName(), n);
    }

    public void recieve(@Nonnull @Const final EventVO vo) {
        String result = null;
        try {
            result = evaluator.eval(vo);
        } catch (Throwable e) {
            LOG.error("Could not evaluate [" + vo + "] with evaluator [" + evaluator + "]", e);
        }

        if (result == Evaluator.ACCEPT) {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    for (Notifier not : notifiers) {
                        try {
                            not.notify(vo);
                        } catch (Throwable e) {
                            LOG.error("Could not notify ["
                                    + vo + "] with notifier ["
                                    + not.getName() + "]", e);
                        }
                    }
                }
            });

        } else if (result != Evaluator.REJECT) {
            final Notifier not = maps.get(result);
            if (not != null) {
                executor.execute(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            not.notify(vo);
                        } catch (Throwable e) {
                            LOG.error("Could not notify ["
                                    + vo + "] with notifier ["
                                    + not.getName() + "]", e);
                        }
                    }
                });

            } else LOG.warn("Could not find notifier named [" + result + "] for event[" + vo + "]");
        }
    }


    public Evaluator getEvaluator() {
        return evaluator;
    }

    public void setEvaluator(Evaluator evaluator) {
        this.evaluator = evaluator;
    }

    public List<Notifier> getNotifiers() {
        return notifiers;
    }

    public void setNotifiers(List<Notifier> notifiers) {
        this.notifiers = new CopyOnWriteArrayList<>(notifiers);
        for (Notifier n : notifiers)
            maps.put(n.getName(), n);
    }

    public void addNotifiers(Notifier notifier) {
        this.notifiers.add(notifier);
    }
}

File: monitor/src/main/scala/com/caibowen/prma/monitor/Monitor.scala
package com.caibowen.prma.monitor

import java.util.{List => JList, Map => JMap}

import akka.AkkaException
import akka.actor._
import akka.actor.SupervisorStrategy._
import com.caibowen.prma.api.model.EventVO
import com.caibowen.prma.monitor.eval._
import com.caibowen.prma.monitor.notify.Notifier
import org.slf4j.LoggerFactory

import scala.util.{Failure, Success, Try}

/**
* @author BowenCai
* @since  30/11/2014.
*/
object Monitor {
  def prop(evaluator: Evaluator, notifierMap: Map[String, Notifier])
   = Props(new Monitor(evaluator, notifierMap))

  final val defaultDecider: SupervisorStrategy.Decider = {
    case _: ActorInitializationException => Stop
    case _: ActorKilledException => Stop
    case _: DeathPactException => Stop
    case _: AkkaException => Restart
    case _: Exception => Escalate
  }
  final val defaultStrategy = new OneForOneStrategy(defaultDecider)
}
@SerialVersionUID(7376528395328272879L)
class Monitor(val evaluator: Evaluator, val notifierMap: Map[String, Notifier]) extends Actor with ActorLogging{

  override val supervisorStrategy = Monitor.defaultStrategy

  private[this] val allNotifiers = notifierMap.values.toList

  def receive = {
    case vo: EventVO => {
      Try(evaluator.eval(vo)) match {
        case Success(response) => response match {
          case n: NotifyOne =>
            val notifier = notifierMap get n.name
            if (notifier.isDefined)
              notifier.get.take(n)
            else log.warning(s"Could not find notifier named [${n.name}] on event [$vo]")
          case n: Response => allNotifiers.foreach(_ take n)
        }

        case Failure(e) => log.error(s"Could not evaluate [$vo] with evaluator [$evaluator]", e)
      }
    }

    case x => unhandled(x)
  }

}
